### ISSUE TITLE
Non-Blitz Land Units Can Move After Capturing a Territory During Non Combat Move (1.9.0.0.6115)

### ISSUE THREAD

**Issue by Cernelius** [triplea-game/triplea#2259]
A "mech_infantry" of WWII Global that captures alone an empty hostile territory with its first move is correctly blocked during Combat Move but it can move on during the subsequent Non Combat Move.
[BugReport_Cernel_20170823_World War II Global 1940 2nd Edition.zip](https://github.com/triplea-game/triplea/files/1243635/BugReport_Cernel_20170823_World.War.II.Global.1940.2nd.Edition.zip)

I believe this bug applies to any maps having multiple movement units unable to blitz.

p.s.: I've never played Global; so, I'd like a Global player to check this is a bug.
p.p.s.: Were there other main referring games before Global having multiple movement units unable to blitz? I believe not.

**Comment by panther2** [comment]
@Cernelius 
I was not able to open your extracted savegame, due to an error when opening.
But I can confirm your issue, having created a respective sitaution with TripleA

No unit - except air units - should be ever allowed to move as well during CM as during NCM.

There is a somehow related note in the document to be found here:
https://github.com/triplea-maps/world_war_ii_global/issues/2

**Comment by Cernelius** [comment]
Actually, I was not talking about that, but about units unable to blitz only.
Meaning that this bug report was not about the limit that "No unit - except air units - should be ever allowed to move as well during CM as during NCM", but about the rule that units unable to bliz should be blocked in the first hostile territory they enter, and stay blocked in there till their endturn.

**I'm fine with the general behaviour that TripleA illegaly allows splitting move between CM and NCM** (for all units, not only air), and surely don't want that limit to be supported (let the players know the rules and follow them), unless having a property for not having it, because I believe it's a nonsensical limit, thus I think custom TripleA maps should be allowed not having to follow such nonsense (as per how it works now).
For example, I'm fine and would want the behaviour being kept that you can move an Armour into a hostile territory during CM and, then, move the same into a friendly territory, during NCM, even tho this is against the rules.
This is because I think it makes totally sense that a unit can move partially in CM and, for the remaining part, in NCM, not only air.
What I'm saying here is only that units unable to blitz should be definitively blocked (until their end-turn phase) upon entering a hostile territory, comprising remaining blocked in the subsequent NCM.
Specifically, a mech_infantry should be forbidden to move into hostile in CM and, then, friendly in NCM, while an armour should be allowed to do the same thing, even tho, by rules, they are both illegal (the difference, rules wise, being that the armour can do it all in CM, while the mech_infantry never (alone)).

So, to sum it up, what I'm asking here is not to enforce the general limit for not-air units, but only solve this bug for not-blitz units:
- Not to enforce the general limit of being unable to split movement between CM and NCM (because it's a dumb rule) or, if enforced, leaving an option for not having it (preferably being able to set what units don't have that limit, default being all air units).
- Just enforce the limit for units unable to blitz that, when they enter any hostile territory, they are blocked in there till their next endturn phase (currently, instead, the engine blocks them for CM, but not for the subsequent NCM, and this is the reported bug).

**Comment by panther2** [comment]
How should I have gotten that from your first posting?

> A "mech_infantry" of WWII Global that captures alone an empty hostile territory with its first move is correctly blocked during Combat Move but it can move on during the subsequent Non Combat Move.

This is simply against the rules, as correctly pointed out, and this is what I confirmed.

In Global the tank is the only unit that can blitz. A Mech Inf is only able to blitz, when paired with a tank.
I don't know how this matters regarding your issue, so I am adding this just for completeness.

Just ignore it in case it is useless.

**Comment by Cernelius** [comment]
And thanks for the confirmation on point 1.
If you (or anyone) can also confirm point 2, that would close the confirmation needs (no doubt this is wrong, but a confirmation is always good to have, in case any developers might need it).

Basically, what the engine is doing here, is for a "mech_infantry" alone having a half-blitz capability, allowing it to blitz only if moving on to friendly territories only (while they should be totally unable to blitz alone).

**Comment by prastle** [comment]
IF all above is correct i agree cer  

What i can say is wow that is a long read ....

I cant confirm or deny since if I noticed it happening I definitely would have said against the rules!

I will trust that u have seen it occur and vote yes

**Comment by panther2** [comment]
> As clarified, that movement is violating two rules:
>
>1. It is violating the general rule that only air unit can split movement between CM and NCM (this rule is just not supposed to be enforced in TA, but left to the players to observe).
>2. It is violating the specific rule that units unable to blitz must definitively stop in the first enemy territory they enter (they are correctly blocked in CM, but not anymore in NCM: fixing this bug would mean keeping them blocked in NCM, until endturn).
>This issue is meant to be about number 2 only. 

Regarding No 1 I would definitely vote for a fix (including a property for disabling it) - at least many A&A players would love to see TripleA being as rules compliant as possible
Regarding No 2 I confirm you are right, the lonely Mech Inf must stop (even in an empty territory), as must a Tank (in case the Tank meets enemy units or enemy infrastructure in that territory)

**Comment by ssoloff** [comment]
Just confirming that the attached save game can't be loaded in the latest pre-releases because it was created with 1.9.0.0.6115, which had a save game compatibility regression.  That regression was fixed in 1.9.0.0.6147 (#2208).

**Comment by Cernelius** [comment]
Just to avoid misunderstandings, I'll clarify that @panther2 with:

> the lonely Mech Inf must stop (even in an empty territory), as must a Tank (in case the Tank meets enemy units or enemy infrastructure in that territory).

means that "the lonely mech_infantry must always stop (even in an empty territory), like an armour would have to in case of meeting enemy units able to block blitzing (any units in WWII Global)".

**Comment by Cernelius** [comment]
So, for this bug, my suggestion is to target the problem related to no-blitz units only, and make sure that the "mech_infantry" (or whatever the like) stays blocked until its EndTurn phase (currently, it is correctly blocked during CM, but stops being blocked at the start of NCM, instead).

Practically, there is something in the engine that un-block the no-blitz unit at the start of NCM, while this should happen only at EndTurn, instead.

I think this is quite important, as it is not easy to pay attention that all no-blitz units were not moved illegally during NCM.

After this is solved, if anyone is interested, the more general matter of forbidding splitting movement between CM and NCM, for not-air units, can be discussed in another issue (I think better waiting till this specific bug for no-blitz units only is saned).

**Comment by ron-murhammer** [comment]
@Cernelius Does this issue occur on other maps like WaW/NWO as well for non-blitz multi-move units?

**Comment by Cernelius** [comment]
It happens in any maps having non-blitz multi-move units, which means not in WAW or NWO, where all multi-move units are able to blitz.

I guess it should be an easy fix. The non-blitz units are already correctly handled during CM. It just needs to avoid NCM to unblock them (endturn should unblock).

**Comment by Cernelius** [comment]
Thinking again about it, instead of EndTurn, it would be probably better that the reset movement unblocks the no-blitz units. I mean this:

>resetUnitStateAtEnd 		= true/false at end of phase will resets movement points and other unit states such as submerged, amphibious, unload/load, wasInCombat, etc. (default at end of phase depends if delegate ends with "NonCombatMove")
>resetUnitStateAtStart 		= true/false at start of phase will resets movement points and other unit states such as submerged, amphibious, unload/load, wasInCombat, etc. (default is false)

I'm thinking it would be logical the most to stay with "wasInCombat" (the one that blocks the units that made combat).

**Comment by Cernelius** [comment]
So, yah, I think default the no-blitz block for having taken undefended hostile territory should be resetted at end of NonCombatMove, or at any point in which "resetUnitState" is true.

**Comment by Cernelius** [comment]
To be clear, in the current engine you can move a ship or an armour of 1 during CM and of another 1 during NCM. This is against the rules, too, but here I'm not asking to fix that (and I actually would not want that to be fixed, unless with a property for not having the fix).

**Comment by panther2** [comment]
> If wanted, another issue may be opened for number 1, tho I believe that should be left not enforced or, if done, I surely suggest having a property for disabling it.

> To be clear, in the current engine you can move a ship or an armour of 1 during CM and of another 1 during NCM. This is against the rules, too, but here I'm not asking to fix that (and I actually would not want that to be fixed, unless with a property for not having the fix).

I will open a separate issue for it, then.

**Comment by Cernelius** [comment]
To me now it reads that they are moving "After Attacking During Non Combat Move", which is not possible, and anyway that is not the matter, as they are correctly blocked if they attacked anything, just not if they conquered or liberated a hostile empty territory, but whatever works best for the ones working on it is good (minor note).

**Comment by ron-murhammer** [comment]
Given that #2266 is a super set of this issue and the fix should address both, I'm closing this one and will track testing in that issue.
